src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.CheckOrReachability.cs

```diff
@@ -0,0 +1,1084 @@
+
+                TypeSymbol narrowedTypeForBinary = NarrowedTypeForBinary(resultLeft, resultRight, binaryPattern.Disjunction);
+
+                return new BoundBinaryPattern(binaryPattern.Syntax, binaryPattern.Disjunction,
+                    left: resultLeft,
+                    right: resultRight,
+                    newInputType, narrowedTypeForBinary);
+            }
+
+            public override BoundNode? VisitPropertySubpattern(BoundPropertySubpattern node)
+            {
+                throw ExceptionUtilities.Unreachable();
+            }
+
+            public override BoundNode? VisitPositionalSubpattern(BoundPositionalSubpattern node)
+            {
+                throw ExceptionUtilities.Unreachable();
+            }
+
+            [return: NotNullIfNotNull(nameof(pattern))]
+            BoundPattern? VisitWithInputTypeFromPattern(BoundPattern? pattern)
+            {
+                if (pattern is null)
+                {
+                    return null;
+                }
+
+                return VisitWithInputType(pattern, pattern.InputType);
+            }
+
+            BoundPattern VisitWithInputType(BoundPattern pattern, TypeSymbol inputType)
+            {
+                var savedInputType = _inputType;
+                _inputType = inputType;
+                var result = (BoundPattern)Visit(pattern);
+                _inputType = savedInputType;
+                return result;
+            }
+
+            private BoundPattern MakeBinary(bool disjunction, ArrayBuilder<BoundPattern> builder, SyntaxNode syntax)
+            {
+                if (builder.Count == 0)
+                {
+                    BoundDiscardPattern discard = MakeDiscardPattern(syntax, _inputType).MakeCompilerGenerated();
+                    if (disjunction)
+                    {
+                        // The caller normally wants to make an `... or ...` sequence but all the parts could be skipped (discards)
+                        // so it can be summed up as a discard too
+                        return discard;
+                    }
+                    else
+                    {
+                        // The caller normally wants to make an `... and ...` sequence but all the parts could be skipped (negated discards)
+                        // so it can be summed up as a negative discard too
+                        return new BoundNegatedPattern(syntax, discard, _inputType, narrowedType: _inputType).MakeCompilerGenerated();
+                    }
+                }
+
+                BoundPattern result;
+                if (disjunction)
+                {
+                    // Produce an `or` chain
+                    result = builder.Last();
+                    Debug.Assert(builder.All(p => p.InputType.Equals(_inputType, TypeCompareKind.AllIgnoreOptions)));
+                    var candidateTypes = ArrayBuilder<TypeSymbol>.GetInstance(2);
+
+                    for (int i = builder.Count - 2; i >= 0; i--)
```

Comment:
>for (int i = builder.Count - 2; i >= 0; i--) [](http://example.com/codeflow?start=20&length=44)

Is there a specific reason to build disjunctions from end? This creates nesting on the right and our implementation is usually prepared to deal with deep nesting on the left.